src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -4552,31 +4589,101 @@ private BoundExpression BindCollectionLiteralExpression(CollectionCreationExpres
+                var builder = new ForEachEnumeratorInfo.Builder();
+                bool hasErrors = !GetEnumeratorInfoAndInferCollectionElementType(syntax, syntax.Expression, ref builder, ref expression, isAsync: false, diagnostics, out _) ||
+                    builder.IsIncomplete;
+                if (hasErrors)
+                {
+                    // PROTOTYPE: Report error.
+                    return BadExpression(syntax);
+                }
 
-                    default:
-                        throw ExceptionUtilities.UnexpectedValue(syntax.Kind());
+                var enumeratorInfo = builder.Build(default);
+                var collectionType = enumeratorInfo.CollectionType;
+                var useSiteInfo = GetNewCompoundUseSiteInfo(diagnostics);
+                var conversion = Conversions.ClassifyConversionFromExpression(expression, collectionType, isChecked: CheckOverflowAtRuntime, ref useSiteInfo);
+                diagnostics.Add(syntax.Expression, useSiteInfo);
+                hasErrors = !conversion.Exists || !conversion.IsImplicit;
+                if (hasErrors)
+                {
+                    GenerateImplicitConversionError(diagnostics, syntax, conversion, expression, collectionType);
+                    // PROTOTYPE:
+                    //// Suppress any additional diagnostics
+                    //diagnostics = BindingDiagnosticBag.Discarded;
+                }
+
+                expression = ConvertForEachCollection(expression, conversion, collectionType, diagnostics);
+                var elementPlaceholder = new BoundValuePlaceholder(syntax.Expression, enumeratorInfo.ElementType);
+                return new BoundCollectionLiteralSpreadElement(
+                    syntax,
+                    expression,
+                    enumeratorInfo,
+                    elementPlaceholder: elementPlaceholder,
+                    elementConversion: null,
+                    addElementPlaceholder: null,
+                    addMethodInvocation: null,
+                    type: enumeratorInfo.CollectionType)
+                { WasCompilerGenerated = true };
+            }
+        }
+
+        private TypeSymbol? InferCollectionLiteralType(ImmutableArray<BoundExpression> initializers)
+        {
+            var useSiteInfo = CompoundUseSiteInfo<AssemblySymbol>.Discarded; // PROTOTYPE: Include use-site diagnostics.
+            // PROTOTYPE: Should IEnumerable contribute object or no type to the best common type?
+            var expressions = initializers.Any(i => i is BoundCollectionLiteralSpreadElement)
+                ? initializers.SelectAsArray(i => i is BoundCollectionLiteralSpreadElement s ? s.ElementPlaceholder : i)
+                : initializers;
+            var bestType = BestTypeInferrer.InferBestType(expressions, Conversions, ref useSiteInfo, out _);
+
+            if (bestType is { } && !bestType.IsVoidType())
```

Comment:
nit: you're welcome to your idioms.  but i personally find the above so baroque, especially compared to the much simpler:

`if (bestType is { SpecialType: not SpecialType.System_Void }`

TO me, that's very clear as to what's going on, emcompasses the null check, doesn't require wondering what the helper does, etc. :)  up to you.

Reply:
We have this same pattern in a number of methods in `Binder`. I'll leave as is, to avoid having this one distinct instance.